### DISCUSSION TITLE
How do you handle settings?

### DISCUSSION THREAD

**Discussion by danechitoaie** [litestar-org/litestar#1548] [Q&A] [answered]
How do you handle loading settings from ENV and then making them available in the route handlers?

At the moment I'm using a BaseSettings class from pydantic. Is this the right way? And how do I make this available in all the route handlers that need it?

**Comment by rgajason** [comment, marked as answer]
We have created an `AppSettings` class based off `BaseSettings`.  We then use dependency injection to make the settings available.

*`config.py`*
```python
class AppSettings(BaseSettings):
    app_name: str = "My Awesome App"
    app_short_name: str = "awesome"
    admin_email: str = "foo@bar.com"
    mongodb_url: str
    mongodb_db: str
```

*`dependencies.py`*
```python
@lru_cache()
def get_settings() -> AppSettings:
    return AppSettings()
```

*`main.py`*
```python
app = Starlite(
    ...
    dependencies={
        "settings": Provide(get_settings, use_cache=True),
    },
)
```

Within a given controller, we can say something like:

```python
async def my_endpoint(
    self,
    settings: AppSettings,
)
```

...and the `settings` variable holds our app settings.

Since this is based off Pydantic's `BaseSettings` we can use a mixture of default / hard-coded, environment variables, and `.env` files.

https://docs.litestar.dev/latest/usage/dependency-injection.html

https://docs.pydantic.dev/usage/settings/

**Reply by danechitoaie** [reply]
What happens with lru_cache if you use Redis caching? will those be cached in Redis and such on every request we will end up with unneeded requests to Redis to retrieve the settings when their values never change?

**Reply by provinzkraut** [reply]
`lru_cache` has nothing to do with Redis. It's a Python stdlib function: https://docs.python.org/3/library/functools.html#functools.lru_cache

**Reply by danechitoaie** [reply]
My bad. Awesome then. I confused this with the caching sport that the framework provides.
Only one more problem that I see with this approach is that the data validation (if the settings are provided or not) that pydantic does happens on request (first time that the dependency is used) vs happening on app startup.

Any way I can do this validation when/or before the app starts and not even start the app if I have some missing configs?

**Reply by provinzkraut** [reply]
You can just do `settings = AppSettings()` anywhere you want and then use that.
